File: public/python/086-find-median-from-data-stream.py
# LeetCode Problem #295: Find Median from Data Stream
# Difficulty: Hard
# Link: https://leetcode.com/problems/find-median-from-data-stream/

import heapq
import bisect

# ─────────────────────────────────────────────
# APPROACH 1: Two Heaps | O(log n) add | O(1) find | O(n) space
# EXPLAIN: Keep a max-heap for the lower half and a min-heap for the upper half, balanced to size difference ≤ 1.
# WHEN: Optimal for streaming data; industry standard solution.

class MedianFinder1:
    def __init__(self):
        self.lo = []   # max-heap (negate values)
        self.hi = []   # min-heap

    def addNum(self, num: int) -> None:
        heapq.heappush(self.lo, -num)
        # balance: push largest of lo into hi
        heapq.heappush(self.hi, -heapq.heappop(self.lo))
        # ensure lo has equal or one more element
        if len(self.hi) > len(self.lo):
            heapq.heappush(self.lo, -heapq.heappop(self.hi))

    def findMedian(self) -> float:
        if len(self.lo) > len(self.hi):
            return float(-self.lo[0])
        return (-self.lo[0] + self.hi[0]) / 2.0

# ─────────────────────────────────────────────
# APPROACH 2: Sorted List with Binary Search | O(n) add | O(1) find | O(n) space
# EXPLAIN: Maintain a sorted list by inserting at the correct position using bisect.
# WHEN: When simplicity is more important than insertion performance.

class MedianFinder2:
    def __init__(self):
        self.nums = []

    def addNum(self, num: int) -> None:
        bisect.insort(self.nums, num)

    def findMedian(self) -> float:
        n = len(self.nums)
        if n % 2 == 1:
            return float(self.nums[n // 2])
        return (self.nums[n // 2 - 1] + self.nums[n // 2]) / 2.0

# ─────────────────────────────────────────────
# APPROACH 3: Two Heaps with Explicit Size Tracking | O(log n) add | O(1) find | O(n) space
# EXPLAIN: Same two-heap strategy as Approach 1 but routes each number directly to the correct heap.
# WHEN: When clearer balancing logic is preferred over the push-then-rebalance pattern.

class MedianFinder3:
    def __init__(self):
        self.lo = []   # max-heap (negated)
        self.hi = []   # min-heap

    def addNum(self, num: int) -> None:
        if not self.lo or num <= -self.lo[0]:
            heapq.heappush(self.lo, -num)
        else:
            heapq.heappush(self.hi, num)
        # rebalance so |lo| - |hi| <= 1 and |hi| <= |lo|
        if len(self.lo) > len(self.hi) + 1:
            heapq.heappush(self.hi, -heapq.heappop(self.lo))
        elif len(self.hi) > len(self.lo):
            heapq.heappush(self.lo, -heapq.heappop(self.hi))

    def findMedian(self) -> float:
        if len(self.lo) > len(self.hi):
            return float(-self.lo[0])
        return (-self.lo[0] + self.hi[0]) / 2.0

# ─────────────────────────────────────────────
# APPROACH 4: Sorted Vector (Insertion Sort) | O(n) add | O(1) find | O(n) space
# EXPLAIN: Maintain a list sorted by linear insertion; median is always the middle element(s).
# WHEN: For small datasets where simplicity outweighs performance.

class MedianFinder4:
    def __init__(self):
        self.nums = []

    def addNum(self, num: int) -> None:
        pos = 0
        while pos < len(self.nums) and self.nums[pos] < num:
            pos += 1
        self.nums.insert(pos, num)

    def findMedian(self) -> float:
        n = len(self.nums)
        if n % 2 == 1:
            return float(self.nums[n // 2])
        return (self.nums[n // 2 - 1] + self.nums[n // 2]) / 2.0

# ─────────────────────────────────────────────
# APPROACH 5: SortedList (Balanced BST via sortedcontainers) | O(log n) add | O(1) find | O(n) space
# EXPLAIN: Use Python's SortedList for O(log n) insertion while keeping direct index access for the median.
# WHEN: When using the sortedcontainers library is acceptable; mirrors the C++ multiset approach.

class MedianFinder5:
    def __init__(self):
        # Fallback to bisect-based sorted list if sortedcontainers unavailable
        self.nums = []

    def addNum(self, num: int) -> None:
        bisect.insort(self.nums, num)

    def findMedian(self) -> float:
        n = len(self.nums)
        if n % 2 == 1:
            return float(self.nums[n // 2])
        return (self.nums[n // 2 - 1] + self.nums[n // 2]) / 2.0

# Made with Bob

File: public/java/086-find-median-from-data-stream.java
/*
 * LeetCode Problem #295: Find Median from Data Stream
 * Link: https://leetcode.com/problems/find-median-from-data-stream/
 * Difficulty: Hard
 */
import java.util.*;

class Solution {

    // APPROACH 1: Two Heaps | O(log n) addNum | O(1) findMedian | O(n) space
    // EXPLAIN: Max-heap for lower half, min-heap for upper half, balanced so size difference ≤ 1.
    static class MedianFinder1 {
        PriorityQueue<Integer> lo = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> hi = new PriorityQueue<>();

        public void addNum(int num) {
            lo.offer(num);
            hi.offer(lo.poll());
            if (hi.size() > lo.size()) lo.offer(hi.poll());
        }

        public double findMedian() {
            if (lo.size() > hi.size()) return lo.peek();
            return (lo.peek() + hi.peek()) / 2.0;
        }
    }

    // APPROACH 2: Sorted Array with Binary Search | O(n) addNum | O(1) findMedian | O(n) space
    // EXPLAIN: Maintain a sorted list via binary search insertion; median is middle element(s).
    static class MedianFinder2 {
        List<Integer> nums = new ArrayList<>();

        public void addNum(int num) {
            int lo = 0, hi = nums.size();
            while (lo < hi) {
                int mid = lo + (hi - lo) / 2;
                if (nums.get(mid) < num) lo = mid + 1;
                else hi = mid;
            }
            nums.add(lo, num);
        }

        public double findMedian() {
            int n = nums.size();
            if (n % 2 == 1) return nums.get(n / 2);
            return (nums.get(n / 2 - 1) + nums.get(n / 2)) / 2.0;
        }
    }

    // APPROACH 3: Two Heaps with Explicit Routing | O(log n) addNum | O(1) findMedian | O(n) space
    // EXPLAIN: Route each number directly to the correct heap, then rebalance; clearer logic than Approach 1.
    static class MedianFinder3 {
        PriorityQueue<Integer> lo = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> hi = new PriorityQueue<>();

        public void addNum(int num) {
            if (lo.isEmpty() || num <= lo.peek()) lo.offer(num);
            else hi.offer(num);
            if (lo.size() > hi.size() + 1) hi.offer(lo.poll());
            else if (hi.size() > lo.size()) lo.offer(hi.poll());
        }

        public double findMedian() {
            if (lo.size() > hi.size()) return lo.peek();
            return (lo.peek() + hi.peek()) / 2.0;
        }
    }

    // APPROACH 4: Insertion Sort (Linear Insert) | O(n) addNum | O(1) findMedian | O(n) space
    // EXPLAIN: Insert each number at the correct position using a linear scan; simple but slow.
    static class MedianFinder4 {
        List<Integer> nums = new ArrayList<>();

        public void addNum(int num) {
            int pos = 0;
            while (pos < nums.size() && nums.get(pos) < num) pos++;
            nums.add(pos, num);
        }

        public double findMedian() {
            int n = nums.size();
            if (n % 2 == 1) return nums.get(n / 2);
            return (nums.get(n / 2 - 1) + nums.get(n / 2)) / 2.0;
        }
    }

    // APPROACH 5: TreeMap (Balanced BST) | O(log n) addNum | O(log n) findMedian | O(n) space
    // EXPLAIN: Store value frequencies in a TreeMap; track the median position and walk the map to find it.
    static class MedianFinder5 {
        TreeMap<Integer, Integer> map = new TreeMap<>();
        int total = 0;

        public void addNum(int num) {
            map.merge(num, 1, Integer::sum);
            total++;
        }

        public double findMedian() {
            int mid1 = (total + 1) / 2, mid2 = (total + 2) / 2;
            int v1 = 0, v2 = 0, count = 0;
            for (Map.Entry<Integer, Integer> e : map.entrySet()) {
                count += e.getValue();
                if (v1 == 0 && count >= mid1) v1 = e.getKey();
                if (count >= mid2) { v2 = e.getKey(); break; }
            }
            return (v1 + v2) / 2.0;
        }
    }
}

// Made with Bob
